modules/rostests/kmtests/ntos_ob/ObInsert.c

```diff
@@ -0,0 +1,651 @@
+    return TRUE;
+}
+
+static
+NTSTATUS
+NTAPI
+QueryNameProc(
+    IN PVOID Object,
+    IN BOOLEAN HasObjectName,
+    OUT POBJECT_NAME_INFORMATION ObjectNameInfo,
+    IN ULONG Length,
+    OUT PULONG ReturnLength,
+    IN KPROCESSOR_MODE AccessMode
+)
+{
+    DPRINT("QueryNameProc() 0x%p, HasObjectName %d, Len %lu, AccessMask 0x%lX\n",
+           Object,
+           HasObjectName,
+           Length,
+           AccessMode);
+
+    ObjectNameInfo = NULL;
+    ReturnLength = 0;
+    return STATUS_SUCCESS;
+}
+
+/*!
+ * @brief This function creates a test directory.
+ */
+static
+VOID
+ObInsert_CreateTestDirectory()
+{
+    NTSTATUS Status;
+
+    /* Initialize the directory name */
+    RtlInitUnicodeString(&DirectoryName, L"\\ObInsertTest");
+
+    /* Initialize object attributes for the directory */
+    InitializeObjectAttributes(&DummyObjectAttributes,
+                               &DirectoryName,
+                               OBJ_KERNEL_HANDLE | OBJ_PERMANENT | OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Create the directory object */
+    Status = ZwCreateDirectoryObject(&DirectoryHandle,
+                                     DELETE,
+                                     &DummyObjectAttributes);
+
+    ok_eq_hex(Status, STATUS_SUCCESS);
+}
+
+/*!
+ * @brief This function creates dummy object type.
+ */
+static
+VOID
+ObInsert_CreateDummyType()
+{
+    NTSTATUS Status;
+
+    /* Initialize the dummy type name */
+    RtlInitUnicodeString(&DummyTypeName, L"ObInsertDummyType");
+
+    RtlZeroMemory(&ObjectTypeInitializer, sizeof(ObjectTypeInitializer));
+
+    ObjectTypeInitializer.PoolType = NonPagedPool;
+    ObjectTypeInitializer.Length = sizeof(ObjectTypeInitializer);
+    ObjectTypeInitializer.ValidAccessMask = OBJECT_TYPE_ALL_ACCESS;
+    ObjectTypeInitializer.MaintainHandleCount = TRUE;
+
+    /* Create the dummy object type (expected to fail initially) */
+    Status = ObCreateObjectType(&DummyTypeName,
+                                &ObjectTypeInitializer,
+                                NULL,
+                                &DummyType);
+
+    ok_eq_hex(Status, STATUS_INVALID_PARAMETER);
+
+    ObjectTypeInitializer.CloseProcedure = CloseProc;
+    ObjectTypeInitializer.DeleteProcedure = DeleteProc;
+    ObjectTypeInitializer.DumpProcedure = DumpProc;
+    ObjectTypeInitializer.OpenProcedure = OpenProc;
+    ObjectTypeInitializer.ParseProcedure = ParseProc;
+    ObjectTypeInitializer.OkayToCloseProcedure = OkayToCloseProc;
+    ObjectTypeInitializer.QueryNameProcedure = QueryNameProc;
+
+    /* Create the dummy object type again with proper procedures */
+    Status = ObCreateObjectType(&DummyTypeName,
+                                &ObjectTypeInitializer,
+                                NULL,
+                                &DummyType);
+
+    /* Handle object name collision status */
+    if (Status == STATUS_OBJECT_NAME_COLLISION)
```

Comment:
Why would there be collision here? I see three cases:
1. Either somebody else created already an object type with the same name --> I would _**strongly**_ discourage you to try modifying the object type object (as I understand you do below), because you wouldn't have any control on who or what that object type really belongs to (i.e. this is going to be bug-prone).
2. Or it was your test that previously ran, but then this means it didn't properly clean up behind itself during the previous iteration --> this is a bug that needs to be fixed.
3. Or ReactOS' implementation of ObCreateObjectType has a bug in which it initially succeeds for the 1st call where you omitted the callback procedures, and then the 2nd call you make here returns a name collision error. Then, the function has to be fixed, and the test would instead expect either `STATUS_SUCCESS`, or if there is another error, the test would have to stop and not try to compensate for the error.

Reply:
The main reason for handling this status in this way is to match the operation of another object manager test, namely [ObType](https://github.com/reactos/reactos/blob/master/modules/rostests/kmtests/ntos_ob/ObType.c). Here is the place where ObType handles collision status in a special way when creating a type: 
https://github.com/reactos/reactos/blob/77d05dd55217553c95ae035045c9695a9abfa912/modules/rostests/kmtests/ntos_ob/ObType.c#L216-L220
That test mentions that type objects are not deleted. Let's take a look at the ObType test's resource cleanup function:
https://github.com/reactos/reactos/blob/77d05dd55217553c95ae035045c9695a9abfa912/modules/rostests/kmtests/ntos_ob/ObType.c#L378-L381
The function does return before deleting object types, leaving behind a piece of dead code. A comment mentions that creating/deleting types makes "Windows unstable".

In my test, I originally decided to delete the created type in the resource cleanup function anyway and create the type again when re-running the test. And this does cause unstable operation of the system, but in my case not Windows, but ReactOS.
1) Either the instability problem is not only on Windows, but also on ReactOS (the behavior is copied).
2) Either the comment in the test has a mistake/typo and instead of Windows, it means ReactOS.
3) These cases are not related.
Point (3) is unlikely in my opinion, because following the ObType test, I removed the recreation of my own type and instead reused an already created type (which is in the final version of this PR at the moment) and the problem disappeared.
Points (1) and (2) should probably be investigated/fixed if necessary by a separate PR (?). In that case I would either:
1) leave it as is with collision handling similar to another Object Manager test.
2) add correct deletion of the created type object despite instability, as in numerous local tests the problem is reproduced on 5-6th test runs and 1st run each time was stable, which may be enough to test regressions?